### PR TITLE
[feat][plugins/aws] Support for Aws SageMaker

### LINKED ISSUE #1176: [AWS] Support SageMaker Service
### Motivation

Collect [AWS SageMaker](https://aws.amazon.com/sagemaker) resources as part of the AWS collect cycle.

### Desired Behavior

Implement the resource collector as part of the [aws](https://github.com/someengineering/resoto/tree/main/plugins/aws) collector. To get an idea how a collector looks like, take the [SQS](https://github.com/someengineering/resoto/blob/main/plugins/aws/resoto_plugin_aws/resource/sqs.py) resource as inspiration. 

For further explanation and guidance, please visit the [Resource Guide](https://resoto.com/docs/edge/contributing/plugins/aws).

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by anjafr**
# Description

New resources:
- `aws_sagemaker_notebook`
- `aws_sagemaker_algorithm`
- `aws_sagemaker_model`
- `aws_sagemaker_app`
- `aws_sagemaker_domain`
- `aws_sagemaker_experiment`
- `aws_sagemaker_trial`
- `aws_sagemaker_project`
- `aws_sagemaker_code_repository`
- `aws_sagemaker_endpoint`
- `aws_sagemaker_image`
- `aws_sagemaker_artifact`
- `aws_sagemaker_user_profile`
- `aws_sagemaker_pipeline`
- `aws_sagemaker_workteam`
- `aws_sagemaker_auto_ml_job`
- `aws_sagemaker_compilation_job`
- `aws_sagemaker_edge_packaging_job`
- `aws_sagemaker_hyper_parameter_tuning_job`
- `aws_sagemaker_inference_recommendations_job`
- `aws_sagemaker_labeling_job`
- `aws_sagemaker_processing_job`
- `aws_sagemaker_training_job`
- `aws_sagemaker_transform_job`

Also introduces:
- `AwsClient.call()`-method without retries and no error handling
- `AwsClient.get_with_retry()` method **with** retries and error handling
- `AwsS3Bucket.name_from_path()` method to get the bucket name from path or uri

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `tox`
- [ ] Document new or updated functionality (someengineering/resoto.com#XXXX)

# Issues Fixed

- Fixes #1176 

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

File: plugins/aws/resoto_plugin_aws/resource/sagemaker.py [hunk 1]
```diff
@@ -0,0 +1,4682 @@
...
+    mapping: ClassVar[Dict[str, Bender]] = {
+        "name": S("Name"),
+        "min_value": S("MinValue"),
+        "max_value": S("MaxValue"),
+        "scaling_type": S("ScalingType"),
+    }
+    name: Optional[str] = field(default=None)
+    min_value: Optional[str] = field(default=None)
+    max_value: Optional[str] = field(default=None)
+    scaling_type: Optional[str] = field(default=None)
```
Comment (aquamatthias) [inline plugins/aws/resoto_plugin_aws/resource/sagemaker.py:2799]: same definition - can be one kind.

File: plugins/aws/resoto_plugin_aws/resource/sagemaker.py [hunk 2]
```diff
@@ -0,0 +1,4682 @@
...
+    mapping: ClassVar[Dict[str, Bender]] = {
+        "auto_generate_endpoint_name": S("AutoGenerateEndpointName"),
+        "endpoint_name": S("EndpointName"),
+    }
+    auto_generate_endpoint_name: Optional[bool] = field(default=None)
+    endpoint_name: Optional[str] = field(default=None)
+
+
+@define(eq=False, slots=False)
+class AwsSagemakerAutoMLJob(AwsResource):
```
Comment (aquamatthias) [inline plugins/aws/resoto_plugin_aws/resource/sagemaker.py:2374]: Would it make sense to define an abstract `AwsSagemakerJob` class? 
A couple of attributes are shared between all jobs.

File: plugins/aws/resoto_plugin_aws/resource/sagemaker.py [hunk 3]
```diff
@@ -0,0 +1,4682 @@
...
+
+
+@define(eq=False, slots=False)
+class AwsSagemakerEnvironmentParameterRanges:
+    kind: ClassVar[str] = "aws_sagemaker_environment_parameter_ranges"
+    mapping: ClassVar[Dict[str, Bender]] = {
+        "categorical_parameter_ranges": S("CategoricalParameterRanges", default=[])
+        >> ForallBend(AwsSagemakerCategoricalParameter.mapping)
+    }
+    categorical_parameter_ranges: List[AwsSagemakerCategoricalParameter] = field(factory=list)
```
Comment (aquamatthias) [inline plugins/aws/resoto_plugin_aws/resource/sagemaker.py:3216]: That's basically a simple list - consider to remove this type

File: plugins/aws/resoto_plugin_aws/resource/sagemaker.py [hunk 4]
```diff
@@ -0,0 +1,4682 @@
...
+    }
+    s3_data_source: Optional[str] = field(default=None)
+    sns_data_source: Optional[str] = field(default=None)
+
+
+@define(eq=False, slots=False)
+class AwsSagemakerLabelingJobDataAttributes:
+    kind: ClassVar[str] = "aws_sagemaker_labeling_job_data_attributes"
+    mapping: ClassVar[Dict[str, Bender]] = {"content_classifiers": S("ContentClassifiers", default=[])}
+    content_classifiers: List[str] = field(factory=list)
```
Comment (aquamatthias) [inline plugins/aws/resoto_plugin_aws/resource/sagemaker.py:3530]: This is a simple list - consider to remove it.

**Comment by anjafr** [comment]
> I saw a couple of properties that are reference ids to other resources. We can drop the id if the edge is semantically the same as having the id (e.g. this is not true if you have multiple relationships for different properties). I only marked some examples - please walk the list of resources if there are more.

@aquamatthias I'm not sure I understand. Do you mean that we should add the edges from the `source` information directly? That's certainly possible, but it would be different to the way we do it for (afaik) every other service so far 🤔

**Comment by aquamatthias** [comment]
> > I saw a couple of properties that are reference ids to other resources. We can drop the id if the edge is semantically the same as having the id (e.g. this is not true if you have multiple relationships for different properties). I only marked some examples - please walk the list of resources if there are more.
> 
> @aquamatthias I'm not sure I understand. Do you mean that we should add the edges from the `source` information directly? That's certainly possible, but it would be different to the way we do it for (afaik) every other service so far 🤔

Yes. See ec2 for examples.

**Comment by anjafr** [comment]
> Yes. See ec2 for examples.

Yea I see 🤔 I'm a bit confused as to why this hasn't come up for past services/resources? I haven't added edges from source anywhere until now (always from the resource model properties) so it would be quite a break from that. And should this then be done retrospectively for all other resources as well? That's quite a big task, so I wonder..

**Comment by anjafr** [comment]
> I also remember that you wanted to introduce an abstract class for all jobs?

So I've been looking at all the job models and it doesn't actually lend itself to an abstract `job` class. The similarities don't hold up. Most jobs have something like Input- and Output-Config for example, but what's inside those differs from job type to job type. Same with VPC-Config - most jobs have it somewhere, but at varying levels of nested information 🤷🏻‍♀️ 

What I originally wanted to have (and still do) is a way to search for instances of any job type at once. Something that would behave as if we had `kind=aws_sagemaker_job` in a search. But I'm not sure how to go about that without overriding the individual kinds of each job 🤔

**Review by aquamatthias — CHANGES_REQUESTED** [review]
Collected AWS sagemaker resources - worked!
I see this error in the logs:
```
ERROR|Duplicate checksum z3wG8LHnb4lcaQ3gT03u9Q for node aws_sagemaker_app default in graph
```

This indicates that the key properties for sagemaker apps are not unique.

I tried to cleanup resources and got those errors:

```
23-01-20 16:33:23|ERROR|An error occurred during clean up aws_sagemaker_app datascience-1-0-ml-t3-medium-ad10824985c0f8abb6eaad94debc in account someengineering (882347060974) region eu-central-1
Traceback (most recent call last):
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/__init__.py", line 343, in cleanup
    if deleted := resource.delete_resource(client):
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/resource/sagemaker.py", line 819, in delete_resource
    client.call(
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 49, in wrapped_f
    return Retrying(*dargs, **dkw).call(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 206, in call
    return attempt.get(self._wrap_exception)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 247, in get
    six.reraise(self.value[0], self.value[1], self.value[2])
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/six.py", line 719, in reraise
    raise value
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 200, in call
    attempt = Attempt(fn(*args, **kwargs), attempt_number, False)
                      ^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/resotolib/resotolib/utils.py", line 154, in timer
    ret = f(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/aws_client.py", line 141, in call
    return self.call_single(aws_service, action, result_name, max_attempts=5, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/aws_client.py", line 119, in call_single
    result = getattr(client, py_action)(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/client.py", line 515, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/client.py", line 893, in _make_api_call
    request_dict = self._convert_to_request_dict(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/client.py", line 964, in _convert_to_request_dict
    request_dict = self._serializer.serialize_to_request(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/validate.py", line 381, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Missing required parameter in input: "UserProfileName"
Unknown parameter in input: "SpaceName", must be one of: DomainId, UserProfileName, AppType, AppName
23-01-20 16:33:23|ERROR|An error occurred during clean up aws_sagemaker_app default in account someengineering (882347060974) region eu-central-1
Traceback (most recent call last):
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/__init__.py", line 343, in cleanup
    if deleted := resource.delete_resource(client):
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/resource/sagemaker.py", line 819, in delete_resource
    client.call(
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 49, in wrapped_f
    return Retrying(*dargs, **dkw).call(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 206, in call
    return attempt.get(self._wrap_exception)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 247, in get
    six.reraise(self.value[0], self.value[1], self.value[2])
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/six.py", line 719, in reraise
    raise value
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/retrying.py", line 200, in call
    attempt = Attempt(fn(*args, **kwargs), attempt_number, False)
                      ^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/resotolib/resotolib/utils.py", line 154, in timer
    ret = f(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/aws_client.py", line 141, in call
    return self.call_single(aws_service, action, result_name, max_attempts=5, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/plugins/aws/resoto_plugin_aws/aws_client.py", line 119, in call_single
    result = getattr(client, py_action)(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/client.py", line 515, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/client.py", line 893, in _make_api_call
    request_dict = self._convert_to_request_dict(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/client.py", line 964, in _convert_to_request_dict
    request_dict = self._serializer.serialize_to_request(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/matthias/Documents/Work/someeng/resoto/venv/lib/python3.11/site-packages/botocore/validate.py", line 381, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Missing required parameter in input: "UserProfileName"
Unknown parameter in input: "SpaceName", must be one of: DomainId, UserProfileName, AppType, AppName
```

Can you take a look?

**Comment by anjafr** [comment]
`Unknown parameter in input: "SpaceName", must be one of: DomainId, UserProfileName, AppType, AppName`
That's funny though. The docs say that you have to either put `UserProfileName` OR `SpaceName`. This error message seems to imply though that `SpaceName` is useless as a parameter for this command?
image

I've now made it so that the deletion command uses `UserProfileName`, if that value is set. And if not, the deletion command is tried with `SpaceName` instead. 🤷🏻‍♀️

**Comment by anjafr** [comment]
> I see this error in the logs:
> 
> ```
> ERROR|Duplicate checksum z3wG8LHnb4lcaQ3gT03u9Q for node aws_sagemaker_app default in graph
> ```
> 
> This indicates that the key properties for sagemaker apps are not unique.

Hm hm hm. Don't fully underststand this one. What exactly does it mean? I just checked if there's any other resource anywhere that has properties of the same time, but there is none?

File: plugins/aws/resoto_plugin_aws/aws_client.py [hunk 5]
```diff
@@ -146,6 +146,20 @@ def call(
...
+        aws_service: str,
+        action: str,
+        result_name: Optional[str],
+        expected_errors: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> JsonElement:
+        try:
+            return self.call_single(aws_service, action, result_name, max_attempts=1, **kwargs)
+        except ClientError:
+            return None
```
Comment (aquamatthias) [inline plugins/aws/resoto_plugin_aws/aws_client.py:161]: why would we swallow client errors?
I would opt to remove the try/except block

**Review by aquamatthias — APPROVED** [review]
One nit - otherwise lgtm. thanks!

**Review by aquamatthias — APPROVED** [review]
🚀 👍

**Pull request merged**